1. TOC {:toc}

Project Title
Whole-brain molecular axes encoding relative spatial location in the human brain

Brief Project Description
There is strong evidence for large-scale genomic gradients helping to organize the spatial distribution of function in the brain, but this topic is very difficult to study in humans. We use PLS to define spatial gradients of gene expression in the adult human brain, and we explore how these gradients emerge of ontogeny and phylogeny, how they relate to topological brain features, how they may influence different traits, and what their relationship is to cell fate.

Project Lead(s)
Jacob W Vogel; Jakob Seidlitz

Faculty Lead(s)
Theodore Satterthwaite; Aaron Alexander-Bloch

Analytic Replicator
Maxwell Bertolero

Collaborators
Konrad Wagstyl, Ross Markello, Casey Paquola, Alexandr Diaz-Papkovic, Thomas Funck, Alan C Evans, Boris Bernhardt, Bratislav Misic

Project Start Date
October 2018

Current Project Status
Drafting manuscript

Dataset
Primary: Allen Human Brain Atlas
Secondary: Brainspan, GTEx, PSYCHEncode Primate/Human

Github repo
https://github.com/PennLINC/Vogel_PLS_Tx-Space

Path to data on filesystem
n/a

Slack Channel
n/a

Trello board
https://trello.com/c/Ut7atlyO/27-pls-gene-expression-project

Google Drive Folder
https://drive.google.com/drive/folders/1fOZRjUR3lyKsENMgG05WNAaJY26l6ham?usp=sharing

Zotero library
Not yet established.

Current work products
OHBM 2020 presentation --> https://www.humanbrainmapping.org/files/2020/ORAL_SESSION_Modeling_and_Analysis_Multivariate_multimodal_analysis.pdf

Code documentation
Every single aspect of preprocessing, sample selection, analysis, plotting, etc has be thoroughly documented and annotated within 11 Jupyter notebooks, which can be found in the Git repo. These can be run in order (most importantly the first 3 notebooks), and all data and code necessary to execute them is avaialble within the repo *or* instructions are given on how to download necessary data from public repositories. Instructions for cloning python environment are provided in the documentation on the github repo. Since all of these analyses are based on public data, reproducing my environment, following the instructions and running the notebook should be sufficient for completely reproducing every aspect of the manuscript.
