---
layout: default
title: Whole-brain molecular axes encoding relative spatial location in the human brain
has_toc: true
nav_order: 1
---

1. TOC {:toc}

**Project Title**

Whole-brain molecular axes encoding relative spatial location in the human brain


**Brief Project Description**

There is strong evidence for large-scale genomic gradients helping to organize the spatial distribution of function in the brain, but this topic is very difficult to study in humans. We use PLS to define spatial gradients of gene expression in the adult human brain, and we explore how these gradients emerge of ontogeny and phylogeny, how they relate to topological brain features, how they may influence different traits, and what their relationship is to cell fate.


**Project Lead(s)**

Jacob W Vogel

**Faculty Lead(s)**

Theodore Satterthwaite; Jakob Seidlitz


**Analytic Replicator**

Maxwell Bertolero


**Collaborators**

Aaron Alexander-Bloch, Konrad Wagstyl, Ross Markello, Adam Pines, Valerie J Sydnor, Alexandr Diaz-Papkovich, Justine Hansen, Alan C Evans, Boris Bernhardt, Bratislav Misic


**Project Start Date**

October 2018


**Current Project Status**

Preprint published; manuscript submitted


**Dataset**

Primary: Allen Human Brain Atlas

Secondary: Brainspan, GTEx, PSYCHEncode Primate/Human


**Github repo**

https://github.com/PennLINC/Vogel_PLS_Tx-Space


**Path to data on filesystem**

n/a


**Slack Channel**

n/a


**Trello board**

https://trello.com/c/Ut7atlyO/27-pls-gene-expression-project


**Google Drive Folder**

https://drive.google.com/drive/folders/1fOZRjUR3lyKsENMgG05WNAaJY26l6ham?usp=sharing


**Zotero library**

None --> Paperpile used


**Current work products**

OHBM 2020 presentation --> https://www.humanbrainmapping.org/files/2020/ORAL_SESSION_Modeling_and_Analysis_Multivariate_multimodal_analysis.pdf

OHBM 2022 presentation --> https://www.humanbrainmapping.org/files/2022/Oral%20Sessions/Imaging_Genetics_Mapping_the_Effects_of_Genetic_and_Transcriptional_Variation_on_the_Brain.pdf

Preprint --> Coming soon...


**Code documentation**

Every single aspect of preprocessing, sample selection, analysis, plotting, etc has be thoroughly documented and annotated within 9 Jupyter notebooks, which can be found in the Git repo. These can be run in order (most importantly the first 3 notebooks), and all data and code necessary to execute them is avaialble within the repo *or* instructions are given on how to download necessary data from public repositories. Instructions for cloning python environment are provided in the documentation on the github repo. Since all of these analyses are based on public data, reproducing my environment, following the instructions and running the notebook should be sufficient for completely reproducing every aspect of the manuscript.
